src/main/auth/o-auth2.ts

```diff
@@ -0,0 +1,149 @@
+import assert from 'assert';
+import { URLSearchParams } from 'url';
+import { Request } from '../request';
+import { RequestAuthorization, AuthRetryConfig } from '../definitions';
+
+export enum OAuth2GrantType {
+    CLIENT_CREDENTIALS = 'client_credentials',
+    REFRESH_TOKEN = 'refresh_token',
+}
+
+export interface OAuth2Params {
+    // required if we want to enforce refreshing..
+    clientId: string;
+    tokenUrl: string;
+
+    // optional
+    clientSecret?: string;
+    refreshToken?: string;
+    accessToken?: string;
+    expiresAt?: number;
+}
+
+export class OAuth2 implements RequestAuthorization, OAuth2Params {
+    retryConfig: AuthRetryConfig;
+
+    clientId: string;
+    tokenUrl: string;
+    clientSecret?: string;
+    refreshToken?: string;
+    accessToken?: string;
+    expiresAt?: number;
+
+    constructor(params: OAuth2Params, retryConfig?: AuthRetryConfig) {
+        this.tokenUrl = params.tokenUrl;
+        this.clientId = params.clientId;
+        this.refreshToken = params.refreshToken;
+        this.accessToken = params.accessToken;
+        this.expiresAt = params.expiresAt;
+        this.clientSecret = params.clientSecret;
+
+        const defaultRetryConfig = {
+            delay: 1000,
+            attempts: 5,
+            statusCodesToRetry: [[401, 401]]
+        };
+
+        this.retryConfig = retryConfig || defaultRetryConfig;
```

Comment:
Good opportunity to use `??` here

Reply:
Aaactually. Probably worth doing `{ ...DEFAULT_RETRY_CONFIG, ...retryConfig }` — then making `AuthRetryConfig` fields required and, finally, making the parameter `Partial<RetryConfig>`. WDYT?

Reply:
Yup!! that sounds great!! also thanks for letting me know about `Partial<T>` 😀